crates/spartan-farmer/src/spartan.rs

```diff
@@ -0,0 +1,168 @@
+#![warn(missing_debug_implementations, missing_docs)]
+//! This is an adaptation of [SLOTH](https://eprint.iacr.org/2015/366) (slow-timed hash function) into a time-asymmetric permutation using a standard CBC block cipher. This code is largely based on the C implementation used in [PySloth](https://github.com/randomchain/pysloth/blob/master/sloth.c) which is the same as used in the paper.
+
+use sloth256_189::cpu;
+use sloth256_189::cuda;
+
+/// Spartan struct used to encode and validate
+#[derive(Debug, Clone)]
+pub struct Spartan {
+    genesis_piece: [u8; 4096],
+}
+
+impl Spartan {
+    /// New instance with 256-bit prime and 4096-byte genesis piece size
+    pub fn new(genesis_piece: [u8; 4096]) -> Self {
+        Spartan { genesis_piece }
+    }
+}
+
+impl Spartan {
+    /// Create an encoding based on genesis piece using provided encoding key hash, nonce and
+    /// desired number of rounds
+    pub fn encode(&self, encoding_key_hash: [u8; 32], nonce: u64, rounds: usize) -> [u8; 4096] {
+        let mut expanded_iv = encoding_key_hash;
+        for (i, &byte) in nonce.to_le_bytes().iter().rev().enumerate() {
+            expanded_iv[32 - i - 1] ^= byte;
+        }
+
+        let mut encoding = self.genesis_piece;
+
+        cpu::encode(&mut encoding, &expanded_iv, rounds).unwrap();
+
+        encoding
+    }
+
+    pub fn batch_encode(
+        &self,
+        piece_amount: usize,
+        pieces: &mut [u8],
+        encoding_key_hash: [u8; 32],
+        nonce_array: &[u64],
+        rounds: usize,
+    ) {
+        // each expanded_iv will be in format [u8; 32], so `piece_amount` expanded_iv's
+        // should consume [u8; 32 * piece_amount] space.
+        let mut expanded_iv_vector: Vec<u8> = Vec::with_capacity(piece_amount * 32);
+        for x in 0..piece_amount {
```

Comment:
This is not C, iterate over the nonces instead of calculating indexes manually, this is what you want in the end.

Reply:
```let mut nonce_iter = nonce_array.iter();
        let mut expanded_iv;
        for _ in 0..piece_amount {
            // same encoding_key_hash will be used for each expanded_iv
            expanded_iv = encoding_key_hash;

            // select the nonce from nonce_array, xor it with the encoding_key_hash
            for (i, &byte) in nonce_iter
                .next()
                .unwrap()
                .to_le_bytes()
                .iter()
                .rev()
                .enumerate()
            {
                expanded_iv[32 - i - 1] ^= byte;
            }

            /* can replace the upper for loop
            nonce_iter
                .next()
                .unwrap()
                .to_le_bytes()
                .iter()
                .rev()
                .enumerate()
                .map(|(i, &byte)| expanded_iv[32 - i - 1] ^= byte);
            */
            expanded_iv_vector.extend(expanded_iv);
        }```
        
        I have made 2 different versions for the inner for loop. The commented out version gets rid of the for loop. Do you want me to follow a similar strategy for eliminating the outer for loop (probably it is possible, haven't tried yet)?

Reply:
You can also `.zip()` one iterator to another avoiding index management for `expanded_iv` in both cases.

Also why `nonce_iter.next()`, why not iterating over nonces directly instead?

Reply:
I'm not iterating over the nonces directly like `nonce_array.iter().next()...` because the upper for loop would reset where I was left, and I would be getting the first nonce only in each iteration. So I decided to create an iterator outside of the for loop, then cast `next()` on it, to not reset it and keep the state.

Trying zip now

Reply:
I'm confused.

What I meant was something like `for nonce in nonce_array.iter_mut() {}`. You don't need to call `.next()` in majority of cases either.

Reply:
I see your point now.

Reply:
how about this:
```rust
        for nonce in nonce_array {
            // same encoding_key_hash will be used for each expanded_iv
            expanded_iv = encoding_key_hash;

            // select the nonce from nonce_array, xor it with the encoding_key_hash
            nonce
                .to_le_bytes()
                .iter()
                .rev()
                .zip(expanded_iv.iter_mut().rev())
                .map(|(nonce_byte, &mut expanded_iv_byte)| *expanded_iv_byte ^= nonce_byte);
            
            expanded_iv_vector.extend(expanded_iv);
        }
```